### PR TITLE
[DC-794] Add Hierarchy View on Chosen Element using Stubbed AJAX API

### PR DESCRIPTION
**Pull request by evansuslovich**
## Summary of changes:

Stubbing data based off [Lou's search screens](https://dspuxteam.invisionapp.com/console/share/BXU5D489SPR/987306077), a function will take in the concept that the user chose to open hierarchy on, and generate a Map representation of the hierarchy. Passing this Map into TreeGrid, a conversion from a Map to Rows will occur, where each concept has several fields, like its depth in the tree, if it's been fetched from the backend already, and if it's state, is it closed, opening, or open. Once this representation is visualized, the selected concept will be highlighted! 

### What
- Changed `initialRows` in TreeGrid to a Map representing a hierarchy, with a function, `hierarchyMapToRows` to convert into Rows
- Changed Props of TreeGrid to expect a domainOptionRoot and a Map<number, Concept[]>
- If a selected concept exists in DomainCriteriaSelector, a stubbed API will return its hierarchy
- Created `SnapshotBuilderConceptNode` type to represent a node in the hierarchy
- Stubbed API `getConceptsHierarchy` that calls `getHierarchyMap` and returns a Map representing a hierarchy

### Why
- Create a nested environment of concepts when hierarchy view is opened
- Create test data, `SnapshotBuilderConceptNode`, and `getHierarchyMap` function to represent backend
- Bold searched concept name to help user identify what they've selected

### Testing strategy
- [ ] 
- Unit Tests!

https://github.com/DataBiosphere/terra-ui/assets/63474660/13f62e89-3a30-4e87-a3f7-9c873974999c

### CODE REVIEW

File: src/components/TreeGrid.ts [hunk 1]
```diff
@@ -46,13 +46,65 @@ const wrapContent =
...
   });
 
+const hierarchyMapToRows = <T extends RowContents>(hierarchyMap: Map<number, T[]>, domainOptionRoot: T): Row<T>[] => {
+  // initialize an array of rows
+  const rows: Row<T>[] = [];
+  const depth = 0;
+
+  const traverseHierarchy = (parent: T, depth: number) => {
+    // push the parent row
+    rows.push({
```
Comment (pshapiro4broad) [inline src/components/TreeGrid.ts]: Not sure what the new UIWG recommendations are for mutability but in the old conventions, mutability was avoided where possible. Could this be written in a non-mutable way, that constructs `rows` instead of modifying it?

File: src/components/TreeGrid.ts [hunk 2]
```diff
@@ -46,13 +46,65 @@ const wrapContent =
...
+  // get the domainOptionRoot's children
+  const domainOptionRootChildren: T[] | undefined = hierarchyMap.get(domainOptionRoot.id);
+
+  // if the domainOptionRoot doesn't have children return an empty array
+  if (!domainOptionRootChildren) {
+    return [];
+  }
+  domainOptionRootChildren.forEach((domainOptionRootChild) => {
+    traverseHierarchy(domainOptionRootChild, depth);
+  });
```
Comment (pshapiro4broad) [inline src/components/TreeGrid.ts]: If you curry on `depth` you can simplify this call a little
```
  const traverseHierarchy = (depth: number) => (parent: T) => {
...

  domainOptionRootChildren.forEach(traverseHierarchy(depth));
```

File: src/components/TreeGrid.ts [hunk 3]
```diff
@@ -46,13 +46,65 @@ const wrapContent =
...
+      contents: parent,
+      depth,
+      isFetched: true,
+      state: 'open',
+    });
+    // if the parent has Children
+    if (parent.hasChildren) {
+      // get the children
+      const children: T[] | undefined = hierarchyMap.get(parent.id);
+      // if children is undefined, return nothing
```
Comment (pshapiro4broad) [inline src/components/TreeGrid.ts]: Maybe add a comment about how this is expected; we only have children for expanded nodes, so we can expect to get nodes that have children but don't have the children in `hierarchyMap`.

File: src/components/TreeGrid.ts (hunk 1 above)
Comment (evansuslovich) [reply src/components/TreeGrid.ts]: Sounds good!

File: src/components/TreeGrid.ts (hunk 3 above)
Comment (evansuslovich) [reply src/components/TreeGrid.ts]: I agree!

File: src/dataset-builder/ConceptSearch.ts [hunk 4]
```diff
@@ -21,8 +21,9 @@ type ConceptSearchProps = {
   readonly onCommit: (selected: Concept[]) => void;
   readonly onOpenHierarchy: (
     domainOption: SnapshotBuilderDomainOption,
-    selected: Concept[],
-    searchText: string
+    cart: Concept[],
+    searchText: string,
+    selectedConcept?: Concept
```
Comment (rjohanek) [inline src/dataset-builder/ConceptSearch.ts]: I think the name selectedConcept is a bit ambiguous as we also use selected to mean in the cart. Maybe this could be called openedConcept?

File: src/dataset-builder/ConceptSetCreator.ts [hunk 5]
```diff
@@ -23,8 +23,18 @@ export const toConceptSet = (concept: Concept): ConceptSet => {
 export const ConceptSetCreator = (props: ConceptSetCreatorProps) => {
   const { onStateChange, dataset, conceptSetUpdater } = props;
   const { snapshotBuilderSettings, id } = dataset;
+  const hierarchyMap = new Map<number, Concept[]>();
+  // create a root for all domainOptions
+  const domainOptionRoot: Concept = {
```
Comment (rjohanek) [inline src/dataset-builder/ConceptSetCreator.ts:28]: I see why this is needed to show the top level domains. And maybe this is the best way to do that! but I'm wondering if you had thought of any other approaches to this problem? Or maybe theres a better place for this root concept to be defined?

File: src/dataset-builder/DatasetBuilderUtils.test.ts [hunk 6]
```diff
@@ -182,6 +187,80 @@ describe('test conversion of DatasetAccessRequest', () => {
...
+    hierarchyMap.set(100, [dummyGetConceptForId(400), dummyGetConceptForId(408)]);
+    expect(getHierarchyMap(408)).toStrictEqual(hierarchyMap);
+  });
+
+  test('fetching hierarchy map of id 100', () => {
+    const hierarchyMap = new Map<number, Concept[]>();
+    hierarchyMap.set(100, []);
+    expect(getHierarchyMap(100)).toStrictEqual(hierarchyMap);
+  });
+});
```
Comment (rjohanek) [inline src/dataset-builder/DatasetBuilderUtils.test.ts:285]: tests look good!

File: src/dataset-builder/DomainCriteriaSelector.ts [hunk 7]
```diff
@@ -59,19 +64,41 @@ export const saveSelected =
...
+  const [hierarchy, setHierarchy] = useState(new Map<number, Concept[]>());
+  const [isHierarchyLoaded, setIsHierarchyLoaded] = useState(false);
   useOnMount(() => {
-    void loadRootConcepts(() => DataRepo().dataset(datasetId).getConcepts(state.domainOption.root));
+    const selectedConcept = state.selectedConcept;
+    if (selectedConcept) {
+      void loadHierarchyConcepts(() => DataRepo().dataset(datasetId).getConceptsHierarchy(selectedConcept));
+    } else {
+      // get me the children of this concept id
+      void loadRootConcepts(() => DataRepo().dataset(datasetId).getConcepts(state.domainOption.root));
```
Comment (rjohanek) [inline src/dataset-builder/DomainCriteriaSelector.ts:77]: should this be rendering the search state if no concept is selected/opened? Would it be something like this?
void loadRootConcepts(() => DataRepo().dataset(datasetId).searchConcepts(state.domainOption.root, ''));

File: src/dataset-builder/DomainCriteriaSelector.ts [hunk 8]
```diff
@@ -59,19 +64,41 @@ export const saveSelected =
...
+    }
   });
-  return rootConcepts.status === 'Ready'
+
+  useEffect(() => {
+    if (rootConcepts.status === 'Ready') {
```
Comment (rjohanek) [inline src/dataset-builder/DomainCriteriaSelector.ts]: so this (case1 ) correlates with: ln 77 void loadRootConcepts(() => DataRepo().dataset(datasetId).getConcepts(state.domainOption.root));

and (case2) else if (hierarchyConcepts.status === 'Ready') correlates with ln 74  void loadHierarchyConcepts(() => DataRepo().dataset(datasetId).getConceptsHierarchy(selectedConcept));

Is that right? Why do we still need to set the hierarchy for case 1? It isn't going to be rendered, right? It's only if a concept is selected/opened that the hierarchy view is displayed. I just want to make sure I understand what is going on here!

File: src/components/TreeGrid.ts [hunk 9]
```diff
@@ -46,13 +46,67 @@ const wrapContent =
...
+
+  return rows;
+};
+
 type TreeGridProps<T extends RowContents> = {
   /** the columns to display */
   readonly columns: Column<T>[];
   /** the initial rows to display */
-  readonly initialRows: T[];
+  readonly initialHierarchy: Map<number, T[]>;
```
Comment (rjohanek) [inline src/components/TreeGrid.ts]: The transformation layer looks great! Design question: Why does it exist within TreeGrid? Does this change affect the functionality of other usages of TreeGrid?

File: src/dataset-builder/ConceptSearch.ts (hunk 4 above)
Comment (evansuslovich) [reply src/dataset-builder/ConceptSearch.ts]: I agree with this!

File: src/components/TreeGrid.ts (hunk 9 above)
Comment (evansuslovich) [reply src/components/TreeGrid.ts]: Great question, I think we should tie this conversation with @s-rubenstein and @pshapiro4broad. I think TreeGrid.ts is mostly going to be used to visualize hierarchies for the exceptions of things like "Select concept sets". It's dangerous to make `readonly initialRows:  T[] | Map<number, T[]> ` because this frustrates TypeScript. Also, I think it's messy and quite ambiguous if we make both initialRows and initialHierarchy optional as: 
`readonly initialRows?: T[]`
`readonly initialHierarchy?: Map<number, T[]>`

File: src/components/TreeGrid.ts [hunk 10]
```diff
@@ -46,13 +46,67 @@ const wrapContent =
...
+        state: 'open',
+      },
+    ];
+
+    // fetch all children of parent
+    const children: T[] | undefined = hierarchyMap.get(parent.id);
+    // if parent doesn't have children, return updatedRows
+    if (!children) {
+      return updatedRows;
+    }
```
Comment (pshapiro4broad) [inline src/components/TreeGrid.ts]: I think you could use `||` here to provide an empty list for the `forEach()` instead of returning early.
```suggestion
    const children: T[] = hierarchyMap.get(parent.id) || [];
```

File: src/dataset-builder/ConceptSelector.ts [hunk 11]
```diff
@@ -66,14 +80,19 @@ export const ConceptSelector = (props: ConceptSelectorProps) => {
                 h(Link, { 'aria-label': `${label} ${concept.id}`, onClick: () => setCart(_.xor(cart, [concept])) }, [
                   icon(iconName, { size: 16 }),
                 ]),
-                div({ style: { marginLeft: 5 } }, [concept.name]),
+                div({ style: { marginLeft: 5 } }, [
+                  openedConceptName === concept.name
+                    ? div({ style: { fontWeight: 600, whiteSpace: 'pre' } }, [concept.name])
```
Comment (pshapiro4broad) [inline src/dataset-builder/ConceptSelector.ts]: Do you need `whitespace: 'pre'` here?

Comment (evansuslovich) [reply src/dataset-builder/ConceptSelector.ts]: Good catch!

File: src/components/TreeGrid.ts (hunk 10 above)
Comment (evansuslovich) [reply src/components/TreeGrid.ts]: This is a great suggestion!!!! Thank you!

File: src/dataset-builder/ConceptSetCreator.ts (hunk 5 above)
Comment (evansuslovich) [reply src/dataset-builder/ConceptSetCreator.ts:28]: Yes, we can most likely create a global root? After our discussion with Phil it seems like it's ok to create these 'fake' roots

File: src/dataset-builder/DomainCriteriaSelector.ts (hunk 8 above)
Comment (evansuslovich) [reply src/dataset-builder/DomainCriteriaSelector.ts]: Yes! We are using the statuses of the AJAX hooks. If rootConcepts.status === 'Ready' this means that `openedConcept` hasn't been passed to DomainCriteriaSelector and the `getConcepts()` ajax hook has been called. In the other case, if hierarchyConcepts.status === 'Ready' this means `openedConcept` has been passed to DomainCriteriaSelector and we are going to get the hierarchy of this selected concept with the `getHierarchy` ajax hook.

File: src/components/TreeGrid.ts [hunk 12]
```diff
@@ -88,8 +118,9 @@ const getRowIndex = <T extends RowContents>(row: Row<T>, rows: Row<T>[]) =>
   _.findIndex((r) => r.contents.id === row.contents.id, rows);
 
 const TreeGridInner = <T extends RowContents>(props: TreeGridPropsInner<T>) => {
-  const { columns, initialRows, getChildren, gridWidth } = props;
-  const [data, setData] = useState(_.map(wrapContent(0), initialRows));
+  // update customWrapper to each props
```
Comment (evansuslovich) [inline src/components/TreeGrid.ts]: delete this line

**Review by rjohanek — COMMENTED** [review]
It seems like items that are selected when on the hierarchy page are not still selected when you navigate back to the search page. Is this the intended functionality? Or do we want to be able to maintain that state across pages like how a concept that is selected on the search page is maintained across pages?

**Comment by evansuslovich** [comment]
> It seems like items that are selected when on the hierarchy page are not still selected when you navigate back to the search page. Is this the intended functionality? Or do we want to be able to maintain that state across pages like how a concept that is selected on the search page is maintained across pages?

Are you referring to items being selected as items in cart? Currently items can continue to be added regardless if its been added to cart.

**Comment by rjohanek** [comment]
> > It seems like items that are selected when on the hierarchy page are not still selected when you navigate back to the search page. Is this the intended functionality? Or do we want to be able to maintain that state across pages like how a concept that is selected on the search page is maintained across pages?
> 
> Are you referring to items being selected as items in cart? Currently items can continue to be added regardless if its been added to cart.

 > Are you referring to items being selected as items in cart? 

Yes. When I select 403 in the hierarchy page it is not still selected when I navigate back to the search page. I'm wondering if this is the intended functionality or if we want to maintain that state.

https://github.com/DataBiosphere/terra-ui/assets/55256690/8705adc8-960d-49da-bde3-8e3c3b1a7046

**Comment by evansuslovich** [comment]
> > > It seems like items that are selected when on the hierarchy page are not still selected when you navigate back to the search page. Is this the intended functionality? Or do we want to be able to maintain that state across pages like how a concept that is selected on the search page is maintained across pages?
> > 
> > 
> > Are you referring to items being selected as items in cart? Currently items can continue to be added regardless if its been added to cart.
> 
> > Are you referring to items being selected as items in cart?
> 
> Yes. When I select 403 in the hierarchy page it is not still selected when I navigate back to the search page. I'm wondering if this is the intended functionality or if we want to maintain that state.
> 
>  Screen.Recording.2024-02-09.at.9.02.15.AM.mov

Thank you for the video! I see now. I've brought something like this up to Phil, where items [In cart continue to show up in search](https://github.com/DataBiosphere/terra-ui/pull/4583#discussion_r1478665761) allowing several instances of the same concept to be added to cart. I personally don't think maintaining selected items in state is the intended functionality but I think this brings up a good parking lot topic :)

File: src/dataset-builder/TestConstants.ts [hunk 13]
```diff
@@ -1,5 +1,10 @@
 import _ from 'lodash/fp';
-import { DatasetModel, SnapshotBuilderConcept as Concept } from 'src/libs/ajax/DataRepo';
+import {
+  DatasetModel,
+  SnapshotBuilderConcept as Concept,
+  SnapshotBuilderConceptNode as ConceptNode,
+} from 'src/libs/ajax/DataRepo';
+// import { RowContents } from "src/components/TreeGrid";
```
Comment (rjohanek) [inline src/dataset-builder/TestConstants.ts]: do we want to remove this import

File: src/dataset-builder/TestConstants.ts [hunk 14]
```diff
@@ -98,19 +103,124 @@ export const dummyConcepts = [
...
+      const children = node.children;
+      const childrenObjects = _.map((childID) => dummyGetConceptForId(childID), children);
+      hierarchyMap.set(node.concept.id, childrenObjects);
+    }
+    if (node.parent) {
+      populateHierarchyMap(dummyGetNodeFromHierarchy(node.parent));
+    }
+  };
+
+  // get the selectedConceptNode
```
Comment (rjohanek) [inline src/dataset-builder/TestConstants.ts]: not necessary, just an idea- selected is a bit ambiguous due to the cart. I see elsewhere you use opened, maybe just in the comment you could clarify what you mean here

**Review by rjohanek — APPROVED** [review]
looks good!!! I'm just about what comes next. Do we want to implement search and hierarchy views for 'select concept sets' as well? If so this is a great precedent!

File: src/dataset-builder/TestConstants.ts (hunk 13 above)
Comment (evansuslovich) [reply src/dataset-builder/TestConstants.ts]: great spot!

File: src/dataset-builder/TestConstants.ts (hunk 14 above)
Comment (evansuslovich) [reply src/dataset-builder/TestConstants.ts]: I'll change to openedConcept! Great Call!

**Comment by evansuslovich** [comment]
> looks good!!! I'm just about what comes next. Do we want to implement search and hierarchy views for 'select concept sets' as well? If so this is a great precedent!

'Select concept sets' should already be implemented :) What comes next ... 
- Replacing stubbed backend call in AJAX api to the backend 
  - Creating backend call 
    - Creating dummyGetNodeFromHierarchy function, is data already in a hierarchy? 
  - Testing backend call
  - Yaml file? 
- Continue to test :) 
- Changing UI to match Lou's screens

File: src/components/TreeGrid.test.ts [hunk 15]
```diff
@@ -2,30 +2,38 @@ import { screen } from '@testing-library/react';
 import userEvent from '@testing-library/user-event';
 import _ from 'lodash/fp';
 import { RowContents, TreeGrid } from 'src/components/TreeGrid';
+import { SnapshotBuilderConcept as Concept } from 'src/libs/ajax/DataRepo';
```
Comment (pshapiro4broad) [inline src/components/TreeGrid.test.ts]: Does this test need to depend on `Concept`? It would be better if it could be written to use `Node` directly instead, since `TreeGrid` itself shouldn't require the use of any snapshot builder types.

File: src/components/TreeGrid.ts [hunk 16]
```diff
@@ -46,13 +46,43 @@ const wrapContent =
...
 type TreeGridProps<T extends RowContents> = {
   /** the columns to display */
   readonly columns: Column<T>[];
   /** the initial rows to display */
-  readonly initialRows: T[];
+  readonly initialHierarchy: Map<number, T[]>;
   /** Given a row, return its children. This is only called if row.hasChildren is true. */
   readonly getChildren: (row: T) => Promise<T[]>;
+  /** Given the domain option root, create a hierarchy */
+  readonly domainOptionRoot: T;
```
Comment (pshapiro4broad) [inline src/components/TreeGrid.ts]: I'd like to keep snapshot builder out of `TreeGrid` if possible. In terms of `TreeGrid`, this is the tree's root.
```suggestion
  /** The root of the tree's hierarchy. */
  readonly root: T;
```

File: src/components/TreeGrid.ts [hunk 17]
```diff
@@ -46,13 +46,43 @@ const wrapContent =
...
+
+  return rows;
+};
+
 type TreeGridProps<T extends RowContents> = {
   /** the columns to display */
   readonly columns: Column<T>[];
   /** the initial rows to display */
-  readonly initialRows: T[];
+  readonly initialHierarchy: Map<number, T[]>;
```
Comment (pshapiro4broad) [inline src/components/TreeGrid.ts]: Given that `Map` is ordered, could we eliminate the `domainOptionRoot` parameter if we require that the root node is the first entry in the map? That seems reasonable to me.

File: src/components/TreeGrid.ts [hunk 18]
```diff
@@ -46,13 +46,43 @@ const wrapContent =
...
+  const rows: Row<T>[] = [];
+  const depth = 0;
+  // get all children of domain option root
+  const domainOptionRootChildren = hierarchyMap.get(domainOptionRoot.id) || [];
+
+  domainOptionRootChildren.forEach((child) => {
+    rows.push(...traverseHierarchy(child, depth, []));
+  });
+
+  return rows;
```
Comment (pshapiro4broad) [inline src/components/TreeGrid.ts]: Is this necessary? I would expect something like this to work:
```suggestion
  return traverseHierarchy(domainOptionRoot, 0, []);
```

As before I'd like to avoid mutability if possible

File: src/components/TreeGrid.test.ts [hunk 19]
```diff
@@ -102,12 +114,21 @@ describe('TreeGrid', () => {
...
+    await user.click(screen.getByLabelText(`expand ${child2.id}`));
+    expect(screen.queryByText(child3.name)).toBeTruthy();
+
+    // collapse
+    await user.click(screen.getByLabelText(`collapse ${child2.id}`));
+    expect(screen.queryByText(child3.name)).toBeFalsy();
+
+    // expand
+    await user.click(screen.getByLabelText(`expand ${child2.id}`));
+    expect(screen.queryByText(child3.name)).toBeTruthy();
```
Comment (pshapiro4broad) [inline src/components/TreeGrid.test.ts]: The original test didn't have these asserts because there is already are tests for this above. The idea is that a test is only testing one thing at a time, if possible, and this test is to test the cache code in `TreeGrid`.

**Review by pshapiro4broad — COMMENTED** [review]
Still looking, here's my comments on `TreeGrid` code

File: src/dataset-builder/CohortEditor.ts [hunk 20]
```diff
@@ -255,7 +255,7 @@ const AddCriteriaSelector: React.FC<AddCriteriaSelectorProps> = (props) => {
       onChange: async (criteriaOption) => {
         if (criteriaOption !== null) {
           if (criteriaOption.value.kind === 'domain') {
-            onStateChange(domainCriteriaSearchState.new(cohort, criteriaGroup, criteriaOption.value));
+            onStateChange(domainCriteriaSearchState.new(cohort, criteriaGroup, criteriaOption.value, [], ''));
```
Comment (pshapiro4broad) [inline src/dataset-builder/CohortEditor.ts]: Why was this change necessary? `new()` has default values for the last two arguments, so they should be able to be omitted.

File: src/dataset-builder/ConceptSearch.ts [hunk 21]
```diff
@@ -21,8 +21,9 @@ type ConceptSearchProps = {
   readonly onCommit: (selected: Concept[]) => void;
   readonly onOpenHierarchy: (
     domainOption: SnapshotBuilderDomainOption,
-    selected: Concept[],
-    searchText: string
+    cart: Concept[],
+    searchText: string,
+    openedConcept?: Concept
```
Comment (pshapiro4broad) [inline src/dataset-builder/ConceptSearch.ts:26]: Maybe `selectedConcept` instead? `open` to me suggests that the node in the tree itself is open, which isn't true for the selected node.
```suggestion
    selectedConcept?: Concept
```

File: src/dataset-builder/ConceptSelector.ts [hunk 22]
```diff
@@ -31,19 +33,31 @@ export const tableHeaderStyle: CSSProperties = {
...
+  const {
+    title,
+    onCancel,
+    onCommit,
+    actionText,
+    datasetId,
+    initialCart,
+    domainOptionRoot,
+    initialHierarchy,
+    openedConceptName,
```
Comment (pshapiro4broad) [inline src/dataset-builder/ConceptSelector.ts]: Could this use `concept.id` instead to match the line? In the off chance that names aren't unique in the concept list we know that the `id` must be unique.

File: src/components/TreeGrid.test.ts (hunk 19 above)
Comment (evansuslovich) [reply src/components/TreeGrid.test.ts]: Yes, I agree.

File: src/components/TreeGrid.ts (hunk 17 above)
Comment (evansuslovich) [reply src/components/TreeGrid.ts]: I've checked and the root node is the last entry in the map. I can still remove the domainOptionRoot though and instead get the last key of the map.

File: src/components/TreeGrid.ts (hunk 18 above)
Comment (evansuslovich) [reply src/components/TreeGrid.ts]: Yes! You're absolutely right!

File: src/components/TreeGrid.test.ts (hunk 15 above)
Comment (evansuslovich) [reply src/components/TreeGrid.test.ts]: Good point!

File: src/components/TreeGrid.ts (hunk 16 above)
Comment (evansuslovich) [reply src/components/TreeGrid.ts]: I've removed :)

File: src/components/TreeGrid.ts [hunk 23]
```diff
@@ -46,13 +46,41 @@ const wrapContent =
...
+      isFetched: children.length > 0,
+      state: children.length > 0 ? 'open' : 'closed',
+    };
+    // recursively traverse hierarchy of all children
+    const childRows: Row<T>[] = children.flatMap((child) => traverseHierarchy(child, depth + 1, []));
+    return [...previousRows, parentRow, ...childRows];
+  };
+
+  // hierarchyMap assumes that the root is the last key
+  const rootID: number | undefined = Array.from(hierarchyMap.keys()).pop();
```
Comment (evansuslovich) [inline src/components/TreeGrid.ts]: Thoughts on calling this domainOptionRootID or rootID?

Comment (evansuslovich) [reply src/components/TreeGrid.ts]: `Array.from(hierarchyMap.keys()).pop();` syntax comes from https://stackoverflow.com/a/30921449/13271387
Thoughts on this?

File: src/dataset-builder/DomainCriteriaSelector.ts [hunk 24]
```diff
@@ -60,19 +65,44 @@ export const saveSelected =
...
+      setHierarchy(new Map<number, Concept[]>([[state.domainOption.root.id, rootConcepts.state.result]]));
+      setIsHierarchyLoaded(true);
+    } else if (hierarchyConcepts.status === 'Ready') {
+      setHierarchy(hierarchyConcepts.state.result);
+      setIsHierarchyLoaded(true);
+    }
+  }, [rootConcepts, hierarchyConcepts, state.domainOption]);
```
Comment (snf2ye) [inline src/dataset-builder/DomainCriteriaSelector.ts]: Nice use of react hooks!

File: src/dataset-builder/ConceptSearch.test.ts [hunk 25]
```diff
@@ -101,11 +101,7 @@ describe('ConceptSearch', () => {
     await user.click(await screen.findByLabelText(`add ${concept.id}`));
     await user.click(screen.getByLabelText(`open hierarchy ${concept.id}`));
     // Assert
-    expect(onOpenHierarchy).toHaveBeenCalledWith(
-      { id: concept.id, category: domainOption.category, root: concept },
-      [concept],
-      ''
-    );
+    expect(onOpenHierarchy).toHaveBeenCalledWith(domainOption, [concept], '', concept);
```
Comment (evansuslovich) [inline src/dataset-builder/ConceptSearch.test.ts:104]: why the extra `concept`, well onOpenHierarchy has an optional parameter `openedConcept`, which is the concept that is being opened. 

These are the arguments for onOpenHierarchy in this scenario
`toHaveBeenCalledWith(domainOption: SnapshotBuilderDomainOption, cart: Concept[], searchText: string, openedConcept?: Concept) `

File: src/libs/ajax/DataRepo.ts [hunk 26]
```diff
@@ -230,6 +239,14 @@ export const DataRepo = (signal?: AbortSignal): DataRepoContract => ({
         ),
       getConcepts: async (parent: SnapshotBuilderConcept): Promise<GetConceptsResponse> =>
         callDataRepo(`repository/v1/datasets/${datasetId}/snapshotBuilder/concepts/${parent.id}`),
+
+      getConceptsHierarchy: async (_concept: SnapshotBuilderConcept) => {
```
Comment (evansuslovich) [inline src/libs/ajax/DataRepo.ts:236]: stubbed ajax api

File: src/libs/ajax/DataRepo.ts [hunk 27]
```diff
@@ -230,6 +239,14 @@ export const DataRepo = (signal?: AbortSignal): DataRepoContract => ({
...
+        await new Promise((resolve) => setTimeout(resolve, 1000));
+        return Promise.resolve({
+          result: getHierarchyMap(_concept.id),
+        });
+      },
+
```
Comment (evansuslovich) [inline src/libs/ajax/DataRepo.ts:242]: stubbed ajax api to get the hierarchy map

File: src/components/TreeGrid.ts [hunk 28]
```diff
@@ -46,13 +46,41 @@ const wrapContent =
...
+    const parentRow: Row<T> = {
+      contents: parent,
+      depth,
+      isFetched: children.length > 0,
+      state: children.length > 0 ? 'open' : 'closed',
+    };
+    // recursively traverse hierarchy of all children
+    const childRows: Row<T>[] = children.flatMap((child) => traverseHierarchy(child, depth + 1, []));
+    return [...previousRows, parentRow, ...childRows];
+  };
```
Comment (snf2ye) [inline src/components/TreeGrid.ts]: What do we gain by having this function nested in the hierarchyMapToRows function? I think it would be more readable to me to be separate.

File: src/components/TreeGrid.ts (hunk 23 above)
Comment (snf2ye) [reply src/components/TreeGrid.ts]: This looks reasonable to me - But we may consider (in the future) making a change on the API to remove the need for this logic by returning more information about the root in the api response.

**Review by snf2ye — APPROVED** [review]

File: src/dataset-builder/ConceptSetCreator.ts (hunk 5 above)
Comment (pshapiro4broad) [reply src/dataset-builder/ConceptSetCreator.ts:28]: I think a fake root is OK for now. At some point we may want to go back and add TreeGrid support for "the root is a node" and "the root is a list of nodes" to make this less confusing. Right now the root node is required but not shown.

File: src/dataset-builder/DatasetBuilderUtils.test.ts [hunk 29]
```diff
@@ -189,6 +194,80 @@ describe('test conversion of DatasetAccessRequest', () => {
...
+   *     |         \
+   *    407       / \
+   *             /   \
+   *           405   406
+   */
+
+  test('fetching hierarchy map of id 400', () => {
+    const hierarchyMap = new Map<number, Concept[]>();
+    hierarchyMap.set(100, [dummyGetConceptForId(400), dummyGetConceptForId(408)]);
+    expect(getHierarchyMap(400)).toStrictEqual(hierarchyMap);
```
Comment (pshapiro4broad) [inline src/dataset-builder/DatasetBuilderUtils.test.ts:218]: It's nice to see all these tests, but `getHierarchyMap()` is a function only used by stub code and will be removed once the API in the service is live, so the tests are not really needed for it.

File: src/dataset-builder/DatasetBuilderUtils.test.ts [hunk 30]
```diff
@@ -189,6 +194,80 @@ describe('test conversion of DatasetAccessRequest', () => {
   });
 });
 
+describe('test gettingHierarchyMap', () => {
+  /**
+   * HIERARCHY REPRESENTATION
```
Comment (pshapiro4broad) [inline src/dataset-builder/DatasetBuilderUtils.test.ts:199]: Maybe this could be repurposed for the TreeGrid tests?

File: src/dataset-builder/DomainCriteriaSearch.ts [hunk 31]
```diff
@@ -26,14 +26,20 @@ export const DomainCriteriaSearch = (props: DomainCriteriaSearchProps) => {
     domainOption: state.domainOption,
     onCancel: () => onStateChange(cohortEditorState.new(state.cohort)),
     onCommit: saveSelected(state, getNextCriteriaIndex, onStateChange),
-    onOpenHierarchy: (domainOption: DomainOption, cart: Concept[], searchText: string) => {
+    onOpenHierarchy: (
+      domainOption: DomainOption,
+      cart: Concept[],
+      searchText: string,
+      openedConcept: Concept = domainOption.root
```
Comment (pshapiro4broad) [inline src/dataset-builder/DomainCriteriaSearch.ts]: Is this correct? When `onOpenHierarchy` is called in `ConceptSearch`, it always passes a parameter here, so this default value will never be used.

File: src/libs/ajax/DataRepo.ts [hunk 32]
```diff
@@ -23,6 +24,13 @@ export type SnapshotBuilderConcept = {
   hasChildren: boolean;
 };
 
+export type SnapshotBuilderConceptNode = {
```
Comment (pshapiro4broad) [inline src/libs/ajax/DataRepo.ts]: I don't think this is an API type, that is, I don't see how the API will be returning this type. If it's not an API type, it should probably go in `dataset-builder-types.ts` which is where the UI types are.

File: src/dataset-builder/DatasetBuilderUtils.ts [hunk 33]
```diff
@@ -148,6 +148,10 @@ export interface GetConceptsResponse {
   result: Concept[];
 }
 
+export interface GetConceptsHierarchyMapResponse {
```
Comment (pshapiro4broad) [inline src/dataset-builder/DatasetBuilderUtils.ts]: I would remove `Map` from this name as that's specific to the type being returned. In general we've been naming these objects like `ApiRequest` and `ApiResponse` where `Api` is the API name so this would be `GetConceptsHierarchyResponse`
```suggestion
export interface GetConceptsHierarchyResponse {
```

File: src/dataset-builder/DomainCriteriaSelector.ts [hunk 34]
```diff
@@ -60,19 +65,44 @@ export const saveSelected =
...
   const { state, onStateChange, datasetId, getNextCriteriaIndex } = props;
+  const [hierarchy, setHierarchy] = useState(new Map<number, Concept[]>());
+  const [isHierarchyLoaded, setIsHierarchyLoaded] = useState(false);
   useOnMount(() => {
-    void loadRootConcepts(() => DataRepo().dataset(datasetId).getConcepts(state.domainOption.root));
+    const openedConcept = state.openedConcept;
+    if (openedConcept) {
+      void loadHierarchyConcepts(() => DataRepo().dataset(datasetId).getConceptsHierarchy(openedConcept));
+    } else {
+      // get me the children of this concept id
```
Comment (pshapiro4broad) [inline src/dataset-builder/DomainCriteriaSelector.ts]: wording suggestion:
```suggestion
      // get the children of this concept
```

File: src/dataset-builder/DomainCriteriaSelector.ts [hunk 35]
```diff
@@ -60,19 +65,44 @@ export const saveSelected =
...
   const [rootConcepts, loadRootConcepts] = useLoadedData<GetConceptsResponse>();
+  const [hierarchyConcepts, loadHierarchyConcepts] = useLoadedData<GetConceptsHierarchyMapResponse>();
   const { state, onStateChange, datasetId, getNextCriteriaIndex } = props;
+  const [hierarchy, setHierarchy] = useState(new Map<number, Concept[]>());
+  const [isHierarchyLoaded, setIsHierarchyLoaded] = useState(false);
   useOnMount(() => {
-    void loadRootConcepts(() => DataRepo().dataset(datasetId).getConcepts(state.domainOption.root));
+    const openedConcept = state.openedConcept;
+    if (openedConcept) {
+      void loadHierarchyConcepts(() => DataRepo().dataset(datasetId).getConceptsHierarchy(openedConcept));
```
Comment (pshapiro4broad) [inline src/dataset-builder/DomainCriteriaSelector.ts]: I think there's a way to write this so that you only have one state that represents both `if` branches, by using `setHierarchy` for both cases. If so it'd be nice to eliminate the additional state objects.

**Review by pshapiro4broad — COMMENTED** [review]
Some comments, will follow up again later

File: src/components/TreeGrid.ts (hunk 28 above)
Comment (evansuslovich) [reply src/components/TreeGrid.ts]: Good point! That way we can test.

File: src/dataset-builder/ConceptSearch.ts (hunk 21 above)
Comment (evansuslovich) [reply src/dataset-builder/ConceptSearch.ts:26]: Good suggestion, it was originally `selectedConcept` but @rjohanek brought up some concerns. @rjohanek @pshapiro4broad I'd like to hear everyones thoughts!

File: src/dataset-builder/DatasetBuilderUtils.test.ts (hunk 29 above)
Comment (evansuslovich) [reply src/dataset-builder/DatasetBuilderUtils.test.ts:218]: Sounds good! While writing `getHierarchyMap()` it was nice to have these tests but I agree. Once we use the API these should be deleted!

File: src/libs/ajax/DataRepo.ts (hunk 32 above)
Comment (evansuslovich) [reply src/libs/ajax/DataRepo.ts]: Sounds good!

File: src/dataset-builder/DomainCriteriaSearch.ts (hunk 31 above)
Comment (evansuslovich) [reply src/dataset-builder/DomainCriteriaSearch.ts]: Fixed!

File: src/dataset-builder/DatasetBuilderUtils.test.ts (hunk 30 above)
Comment (evansuslovich) [reply src/dataset-builder/DatasetBuilderUtils.test.ts:199]: agreed!

**Review by pshapiro4broad — APPROVED** [review]
Looks good, nice to this this all coming together!
